### ISSUE TITLE
Space weather causes lag (uses all the CPU)

### ISSUE THREAD

**Issue by yjhn** [endless-sky/endless-sky#5644]
**Describe the bug**
In Ember Waste magnetic storms are causing the CPU load to become > 100% if there are many ships. This causes severe lag.

**To Reproduce**
1. Go to Ember Waste with as many ships as possible
2. Wait for magnetic storm to come
3. Observe CPU usage

**Expected behavior**
Space weather does not cause noticeable lag

**System (please complete the following information):**
 - OS: Manjaro Linux
 - Game Source: self-compiled
 - Version: e0ae8bf871f4ec957075e052cb1d0976d5a0de63

**Link to save file**
[Stuck.Remnant.txt](https://github.com/endless-sky/endless-sky/files/5824682/Stuck.Remnant.txt)

**Additional context**
Maybe the effects could be drawn using shaders? Since shaders are executed on GPU, it should be much faster.

**Comment by Zitchas** [comment]
Just out of curiosity, how many ships are you talking about here? A dozen? A hundred? A thousand?

In any event, having the storms displayed via a shader rather than the current method sounds like a good idea, although I am not a coder, so I don't really know what that might entail.

**Comment by Amazinite** [comment]
I imagine that it's the number of effects generated, not the number of ships impacted, that it causing the CPU usage. As far as generating the effects in shaders goes, I have no idea how we'd go about that. As for if anything should change at all with regards to hazards generating effects, my first idea would be to tile environmental effects of hazards like asteroids are tiled, while still limiting the range of where effects appear to the circle or ring that encompasses the hazard, but I'm not actually sure how much that would help given that we'd still need to go through the step of generating effects, which is what is causing the issues. (I suppose the way it would help is by requiring less effects to be generated for the same number of visuals to appear.)

Important to note that any "environmental effect" of a hazard is generated every frame, with the number being affected by the strength of the hazard. So in the case of the ion hazards, we're generating anywhere from 20 to 600 effects per frame. So the obvious approach would be to just limit the number of hazards generated every frame.
In 66607582b22a1a3f5755d7743d5b77f44426b3f5 I've dropped the base effect count from 200 to 20, while at the same time reducing the range of the effect and increasing the lifetime of each effect. This means we should be getting about the same density of visual effects with far less effects needing to be generated.

Given that it is the number of effects that are causing the issue, I'm not sure if any change is necessary. A large number of effects is always going to lag the game no matter where it's coming from. A ship with lots of explosion effects on death for example will lag your game out when those effects are generated. So I think we should just keep in mind to have the number of effects generated by hazard be on the low end, not the several hundred that we started with.

The only thing I can think to do aside from just keeping the number of effects low might be to add an "effect period" attribute to hazards. As mentioned, right now hazards generated their effects every frame they are active. An "effect period" attribute that defaults to 1 would retain that behavior, but if you were to set "effect period" to 2 then hazards would only be generated on average every other frame, on average every third frame for a value of 3, and so on for higher effect period values.

**Comment by Zitchas** [comment]
Given that the player can't really see things happening in 1/60th of a second, altering the effect to have quadruple the effect but only trigger every 4 frames should have the same actual impact on the game while having a quarter the burden on the CPU.

**Comment by Amazinite** [comment]
> altering the effect to have quadruple the effect but only trigger every 4 frames should have the same actual impact on the game while having a quarter the burden on the CPU.

Not really. If the number of effects over a given amount of time is the same then the burden is the same. Four times the effects every four frames is just burdening your CPU a lot every fourth frame instead of a fourth that amount every frame.

**Comment by tatami4** [comment]
> Just out of curiosity, how many ships are you talking about here? A dozen? A hundred? A thousand?
<50 ships

With commit 6660758 it's much better (CPU usage as reported by game does not go above 100%).

Another thing: when the magnetic storm is happening, the game appears to severely slow down.

**Comment by Amazinite** [comment]
I'm not experiencing any sort of slow down on my laptop. Do you have fast forward enabled? If GPU or CPU usage numbers reach over 33% then fast forward will become slowed down.

**Comment by tatami4** [comment]
Yes, fast forward is enabled. Without it the slowdown is far less noticeable.

**Comment by TomGoodIdea** [comment]
Is there anything else we can do to improve the situation? Restricting the simulation to a portion of the system that the player sees would be a breaking change potentially affecting NPC combat and tbh I'm not sure if it's worth the performance with fast-forward enabled.
